File: backend/routes/system.js
/**
 * ClockMQTT — System Routes
 * ==========================
 * System status, auth, health, and device commands.
 *
 * Mounted at /api — handles:
 *   GET  /api/health
 *   GET  /api/system/status
 *   POST /api/auth/login
 *   POST /api/devices/:id/command
 */

const { Router } = require('express');
const mqttClient = require('../mqttClient');
const deviceManager = require('../deviceManager');
const wordStore = require('../wordStore');
const scheduleStore = require('../scheduleStore');

const router = Router();

// ===========================================================================
// GET /api/health
// ===========================================================================

router.get('/health', (req, res) => {
  res.json({
    status: 'ok',
    mqtt_connected: mqttClient.isConnected,
  });
});

// ===========================================================================
// GET /api/system/status — dashboard overview
// ===========================================================================

router.get('/system/status', (req, res) => {
  const stats = deviceManager.getStats();
  res.json({
    mqtt_connected: mqttClient.isConnected,
    devices_online: stats.online,
    devices_total: stats.total,        // FIX: uses registry.size, not max(1, statusCount)
    words_total: wordStore.count(),
    schedules_active: scheduleStore.countActive(),
  });
});

// ===========================================================================
// POST /api/auth/login
// ===========================================================================

router.post('/auth/login', (req, res) => {
  const { username, password } = req.body;
  if (username === 'admin' && password === 'admin123') {
    return res.json({
      status: 'ok',
      token: 'demo-jwt-token-change-in-production',
    });
  }
  res.status(401).json({ detail: 'Invalid credentials' });
});

// ===========================================================================
// POST /api/devices/:id/command — send display text to a device
// ===========================================================================

router.post('/devices/:id/command', (req, res) => {
  try {
    const { id } = req.params;
    const { lines, duration_sec } = req.body;

    if (!lines || !Array.isArray(lines) || lines.length === 0) {
      return res.status(400).json({ detail: 'lines must be a non-empty array' });
    }

    const formattedLines = lines.map(line => ({
      text: line.text || '',
      size: line.size || 32,
      y: line.y || 20,
    }));

    mqttClient.publishDisplayText(id, formattedLines, duration_sec || 30);

    res.json({ status: 'ok', device: id });
  } catch (e) {
    res.status(400).json({ detail: e.message });
  }
});

module.exports = router;

File: backend/routes/system.py
"""System API — status, login."""

import time
from datetime import datetime, timezone, timedelta

from fastapi import APIRouter, HTTPException
from models.schemas import DisplayPushRequest, LoginRequest, SystemStatus
from mqtt_client import mqtt_broker
from routes.devices import _device_status

router = APIRouter(prefix="/api", tags=["system"])

# Timezone for China Standard Time (UTC+8)
CST = timezone(timedelta(hours=8))


@router.get("/system/status")
def get_status():
    """Get system overview."""
    online_count = sum(1 for s in _device_status.values() if s.get("online"))
    return {
        "mqtt_connected": mqtt_broker.is_connected,
        "devices_online": online_count,
        "devices_total": max(1, len(_device_status)),
        "words_total": 3,
        "schedules_active": 0,
    }


@router.post("/auth/login")
def login(req: LoginRequest):
    """Simple admin login."""
    if req.username == "admin" and req.password == "admin123":
        return {"status": "ok", "token": "demo-jwt-token-change-in-production"}
    raise HTTPException(401, "Invalid credentials")


@router.post("/devices/{device_id}/command")
def send_display_command(device_id: str, req: DisplayPushRequest):
    """Send a display text command to a device."""
    lines = [{"text": line.text, "size": line.size, "y": line.y} for line in req.lines]
    mqtt_broker.publish_display_text(device_id, lines, req.duration_sec)
    return {"status": "ok", "device": device_id}
